### ISSUE TITLE
Apply some idiomatic shortcuts

### ISSUE THREAD

**Issue by frol** [Wojtek242/rwmstatus#1]
https://github.com/Wojtek242/rwmstatus/blob/676830f21178041748fff16340726f027cd8d696/src/main.rs#L42

I would use `display.is_null()`

https://github.com/Wojtek242/rwmstatus/blob/676830f21178041748fff16340726f027cd8d696/src/lib.rs#L30-L35

There is a built-in [`std::fs::read_to_string`](https://doc.rust-lang.org/std/fs/fn.read_to_string.html)

Here are a few shortcuts:

https://github.com/Wojtek242/rwmstatus/blob/676830f21178041748fff16340726f027cd8d696/src/lib.rs#L68-L75

```rust
if let Ok(contents) = read_file(&batt, "present") {
    if !contents.starts_with('1') {
        return "not present".into();
    }
} else {
    return "".into();
}
```

https://github.com/Wojtek242/rwmstatus/blob/676830f21178041748fff16340726f027cd8d696/src/lib.rs#L77-L85

```rust
read_file(&batt, "charge_full_design")
    .or_else(|_| read_file(&batt, "energy_full_design"))
    .unwrap_or_else(|| "".into());
```

https://github.com/Wojtek242/rwmstatus/blob/676830f21178041748fff16340726f027cd8d696/src/lib.rs#L87-L90

```rust
let desired_capacity: u64 = co.trim().parse().map_err(|_| "invalid".into())?;
```

**Comment by Wojtek242** [comment]
Thanks for taking the time to read the code and provide these pointers! I found them very helpful.

I have now applied them to the code. There are some differences with what you suggested as I changed how I handle the errors.

I will keep the issue open for some time if you want to have another look at some point.

**Comment by frol** [comment]
https://github.com/Wojtek242/rwmstatus/blob/bd8df56616488cf22df057aa04765b33474f79bd/src/lib.rs#L128-L136

```rust
let dir_filtered = dir.filter(|path_result| {
    path_result
        .ok()
        .and_then(|path| path.file_name().to_str())
        .filter(|entry| entry.starts_with(prefix))
        .is_ok()
});
```

https://github.com/Wojtek242/rwmstatus/blob/bd8df56616488cf22df057aa04765b33474f79bd/src/lib.rs#L138-L143

```rust
let mut paths: Vec<PathBuf> = dir_filtered
    .map(|path_result| path_result.expect("Unexpected file path").path())
    .collect();
```

I would also move 

https://github.com/Wojtek242/rwmstatus/blob/a6d36906a8c78faf557f13c20f2c53cbbb0f275b/src/main.rs#L43

out of the loop and only call `stats.clear()` at the beginning of the loop.

[It also better to use `_else` variants of `.unwrap_or*` family](https://rust-lang-nursery.github.io/rust-clippy/master/index.html#or_fun_call) to avoid unnecessary allocations.

`format!("")` is equivalent to `"".into()`.

**Comment by frol** [comment]
By the way, use [`clippy`](https://github.com/rust-lang/rust-clippy#usage) to learn best practices and [`fmt`](https://github.com/rust-lang/rustfmt#quick-start) to keep your code style consistent.

**Comment by Wojtek242** [comment]
Thanks for the further comments! I've applied all your new suggestions.

However, your suggestion to fix `get_paths` didn't work like that as the borrow checker wasn't happy with it. As I understand it, it was because `ok()` and `and_then()` move the value within the enum which `filter` does not like. Also, `file_name()` returns a reference which doesn't live past the end of the `and_then` in which it is used.

In the end, by rewriting the two blocks slightly differently, I did get it to work using `Result`/`Option` methods instead of matches. So from this

https://github.com/Wojtek242/rwmstatus/blob/a6d36906a8c78faf557f13c20f2c53cbbb0f275b/src/lib.rs#L128-L143

I went to this

https://github.com/Wojtek242/rwmstatus/blob/ce2fe133d5b4818c7d3c894b35a1c919e1bbce65/src/lib.rs#L128-L139

Here, `file_name()` is called first and thus all the subsequent calls only move the reference and there no longer is a lifetime issue. This also has the advantage of eliminating the `panic!` branch (which was hopefully unhittable anyway).

I do use `fmt`, though I use an old version, since the newest one requires you to be on the nightly channel, which I do not want to use.

And I haven't been using `clippy` as I prefer not to use `rustup` and clippy is easiest to install with `rustup`. I prefer to stick to the distribution provided toolchain. However, I did run clippy on the codebase in a VM in the end anyway and will try to do so on occasion.
